### PR TITLE
Fruit shop generator report

### PR DESCRIPTION
**Pull request by Daniel1004K**

### CODE REVIEW

File: src/main/java/core/basesyntax/service/FileCsvWriter.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+public class FileCsvWriter implements Writer {
+    @Override
+    public boolean write(String report, String path) {
+        try {
+            if (!Files.exists(Path.of(path))) {
+                Files.createFile(Path.of(path));
+            }
+            Files.write(Path.of(path),report.getBytes(StandardCharsets.UTF_8));
+        } catch (IOException e) {
+            throw new RuntimeException("There is a problem with saving", e);
```
Comment (Veronika-Alekseienko) [inline src/main/java/core/basesyntax/service/FileCsvWriter.java:17]:
```suggestion
            throw new RuntimeException("There is a problem with saving" + filePath, e);
```

File: src/main/java/core/basesyntax/service/FileCsvWriter.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
+package core.basesyntax.service;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+public class FileCsvWriter implements Writer {
+    @Override
+    public boolean write(String report, String path) {
```
Comment (Veronika-Alekseienko) [inline src/main/java/core/basesyntax/service/FileCsvWriter.java:10]:
```suggestion
    public boolean write(String report, String filePath) {
```

File: src/main/java/core/basesyntax/service/Parser.java [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package core.basesyntax.service;
+
+import core.basesyntax.dto.FruitDto;
+
+public interface Parser {
+    FruitDto parse(String element);
```
Comment (Veronika-Alekseienko) [inline src/main/java/core/basesyntax/service/Parser.java:6]:
```suggestion
    FruitDto parse(String line);
```

File: src/main/java/core/basesyntax/service/ParserCsv.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    public ParserCsv(Validator validator) {
+        this.validator = validator;
+    }
+
+    @Override
+    public FruitDto parse(String element) {
+        if (!validator.test(element)) {
+            throw new RuntimeException("Exception: wrong line");
+        }
+        String[] partsOfElement = element.split(",");
```
Comment (Veronika-Alekseienko) [inline src/main/java/core/basesyntax/service/ParserCsv.java:20]: make constant for ","

File: src/main/java/core/basesyntax/service/Reader.java [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package core.basesyntax.service;
+
+import java.util.List;
+
+public interface Reader {
+    List<String> read(String fileName);
```
Comment (Veronika-Alekseienko) [inline src/main/java/core/basesyntax/service/Reader.java:6]:
```suggestion
    List<String> readFromFile(String fileName);
```

File: src/main/java/core/basesyntax/service/Writer.java [hunk 6]
```diff
@@ -0,0 +1,5 @@
+package core.basesyntax.service;
+
+public interface Writer {
+    boolean write(String report, String path);
```
Comment (Veronika-Alekseienko) [inline src/main/java/core/basesyntax/service/Writer.java:4]:
```suggestion
    boolean write(String report, String filePath);
```

File: src/main/java/core/basesyntax/srategy/PurchaseHandler.java [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+    public int apply(FruitDto fruitDto) {
+        if (!Storage.storage.containsKey(new Fruit(fruitDto.getName()))) {
+            throw new RuntimeException(fruitDto.getName() + "isn't exist");
+        }
+        int count = Storage.storage.get(new Fruit(fruitDto.getName()));
+        if (count - fruitDto.getCount() < 0) {
+            throw new RuntimeException(fruitDto.getName() + "is not enough");
+        }
+        Storage.storage.put(new Fruit(fruitDto.getName()),count - fruitDto.getCount());
+        return count - fruitDto.getCount();
```
Comment (Veronika-Alekseienko) [inline src/main/java/core/basesyntax/srategy/PurchaseHandler.java:18]: let's make variable for    count - fruitDto.getCount()

File: src/main/java/core/basesyntax/srategy/SupplyHandler.java [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+
+public class SupplyHandler implements OperationHandler {
+    @Override
+    public int apply(FruitDto fruitDto) {
+        if (!Storage.storage.containsKey(new Fruit(fruitDto.getName()))) {
+            throw new RuntimeException(fruitDto.getName() + "isn't exist");
+        }
+        int count = Storage.storage.get(new Fruit(fruitDto.getName()));
+        Storage.storage.put(new Fruit(fruitDto.getName()), fruitDto.getCount() + count);
+        return count + fruitDto.getCount();
```
Comment (Veronika-Alekseienko) [inline src/main/java/core/basesyntax/srategy/SupplyHandler.java:15]: let's make variable for   count + fruitDto.getCount()

**Review by Veronika-Alekseienko — CHANGES_REQUESTED** [review]
Well done !!)

**Review by Veronika-Alekseienko — APPROVED** [review]

File: .gitignore [hunk 9]
```diff
@@ -1,3 +1,4 @@
 .idea/*
 *.iml
 target/*
+/src/main/resources/
```
Comment (KseniiaMakarova) [inline .gitignore:4]: let's remove this line) we already pushed these files to PR, and it is ok

File: src/main/java/Main.java [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+import core.basesyntax.srategy.OperationHandler;
+import core.basesyntax.srategy.PurchaseHandler;
+import core.basesyntax.srategy.ReturnHandler;
+import core.basesyntax.srategy.SupplyHandler;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class Main {
+    private static final String STORAGE_PATH = "src/main/resources/storage.csv";
```
Comment (KseniiaMakarova) [inline src/main/java/Main.java:20]:
```suggestion
    private static final String PATH_INPUT_FILE = "src/main/resources/storage.csv";
```

File: src/main/java/core/basesyntax/dto/FruitDto.java [hunk 11]
```diff
@@ -0,0 +1,65 @@
+package core.basesyntax.dto;
+
+import java.util.Objects;
+
+public class FruitDto {
+    private String operation;
+    private String name;
+    private int count;
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/dto/FruitDto.java:8]:
```suggestion
    private int quantity;
```

File: src/main/java/core/basesyntax/service/FileCsvReader.java [hunk 12]
```diff
@@ -0,0 +1,22 @@
+package core.basesyntax.service;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class FileCsvReader implements Reader {
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/service/FileCsvReader.java:9]:
```suggestion
public class CsvFileReader implements FileReader {
```

File: src/main/java/core/basesyntax/service/FileCsvReader.java [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+    @Override
+    public List<String> readFromFile(String fileName) {
+        try {
+            return Files.readAllLines(Paths.get(fileName))
+                .stream()
+                .skip(1)
+                .collect(Collectors.toList());
+        } catch (IOException e) {
+            throw new RuntimeException("Problem with file",e);
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/service/FileCsvReader.java:18]: let's pass fileName to the exception

File: src/main/java/core/basesyntax/service/FileCsvWriter.java [hunk 14]
```diff
@@ -0,0 +1,22 @@
+package core.basesyntax.service;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+public class FileCsvWriter implements Writer {
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/service/FileCsvWriter.java:8]:
```suggestion
public class FileWriterImpl implements FileWriter {
```
we can actually write any content to any file, not only csv

File: src/main/java/core/basesyntax/service/FileCsvWriter.java [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+    @Override
+    public boolean writeToFile(String report, String filePath) {
+        try {
+            if (!Files.exists(Path.of(filePath))) {
+                Files.createFile(Path.of(filePath));
+            }
+            Files.write(Path.of(filePath),report.getBytes(StandardCharsets.UTF_8));
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/service/FileCsvWriter.java:15]: does Files.write already create a file if needed?

File: src/main/java/core/basesyntax/service/FruitReport.java [hunk 16]
```diff
@@ -0,0 +1,25 @@
+package core.basesyntax.service;
+
+import core.basesyntax.db.Storage;
+import core.basesyntax.model.Fruit;
+import java.util.Map;
+
+public class FruitReport {
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/service/FruitReport.java:7]:
```suggestion
public class FruitReportService implements ReportService {
```

File: src/main/java/core/basesyntax/service/FruitReport.java [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+    public static final String FILE_CSV_TITLE = "fruit,quantity\n";
+    private static final char CSV_SEPARATOR = ',';
+    private final Map<Fruit, Integer> storage;
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/service/FruitReport.java:10]: I think we can better use this as local variable

File: src/main/java/core/basesyntax/service/ParserCsv.java [hunk 18]
```diff
@@ -0,0 +1,26 @@
+package core.basesyntax.service;
+
+import core.basesyntax.dto.FruitDto;
+
+public class ParserCsv implements Parser {
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/service/ParserCsv.java:5]:
```suggestion
public class ParserImpl implements Parser {
```

File: src/main/java/core/basesyntax/service/ValidatorCsv.java [hunk 19]
```diff
@@ -0,0 +1,10 @@
+package core.basesyntax.service;
+
+public class ValidatorCsv implements Validator {
+    public static final String PATTERN_ELEMENT_CSV = "[bspr],\\w+,\\d+";
+
+    @Override
+    public boolean test(String o) {
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/service/ValidatorCsv.java:7]: what is `o`?

File: src/main/java/core/basesyntax/srategy/SupplyHandler.java [hunk 20]
```diff
@@ -0,0 +1,19 @@
...
+    public int apply(FruitDto fruitDto) {
+        Fruit fruit = new Fruit(fruitDto.getName());
+        if (!Storage.storage.containsKey(fruit)) {
+            throw new RuntimeException(fruit.getName() + "isn't exist");
+        }
+        int count = Storage.storage.get(fruit);
+        int result = fruitDto.getCount() + count;
+        Storage.storage.put(fruit, result);
+        return result;
+    }
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/srategy/SupplyHandler.java:18]: optional: we can use 1 handler both for supply and return

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/service/CsvFileWriter.java [hunk 21]
```diff
@@ -5,13 +5,10 @@
 import java.nio.file.Files;
 import java.nio.file.Path;
 
-public class FileCsvWriter implements Writer {
+public class CsvFileWriter implements FileWriter {
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/service/CsvFileWriter.java:8]: >we can actually write any content to any file, not only csv

what about this comment?

File: src/main/java/core/basesyntax/service/FruitReportService.java [hunk 22]
```diff
@@ -4,16 +4,15 @@
...
-public class FruitReport {
+public class FruitReportService implements ReportService {
     public static final String FILE_CSV_TITLE = "fruit,quantity\n";
     private static final char CSV_SEPARATOR = ',';
-    private final Map<Fruit, Integer> storage;
 
-    public FruitReport() {
-        this.storage = Storage.storage;
+    public FruitReportService() {
     }
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/service/FruitReportService.java]: do we need the empty constructor?

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request closed without merging**